// sigma/utils/dist/sigma-utils.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var normalization = require('../../dist/normalization-aed467cc.cjs.prod.js');
var colors = require('../../dist/colors-ee2e2828.cjs.prod.js');
var data = require('../../dist/data-881001cd.cjs.prod.js');
require('graphology-utils/is-graph');



exports.ANIMATE_DEFAULTS = normalization.ANIMATE_DEFAULTS;
exports.animateNodes = normalization.animateNodes;
exports.createElement = normalization.createElement;
exports.createNormalizationFunction = normalization.createNormalizationFunction;
exports.cubicIn = normalization.cubicIn;
exports.cubicInOut = normalization.cubicInOut;
exports.cubicOut = normalization.cubicOut;
exports.easings = normalization.easings;
exports.getCorrectionRatio = normalization.getCorrectionRatio;
exports.getMatrixImpact = normalization.getMatrixImpact;
exports.getPixelRatio = normalization.getPixelRatio;
exports.graphExtent = normalization.graphExtent;
exports.identity = normalization.identity;
exports.linear = normalization.linear;
exports.matrixFromCamera = normalization.matrixFromCamera;
exports.multiply = normalization.multiply;
exports.multiplyVec2 = normalization.multiplyVec2;
exports.quadraticIn = normalization.quadraticIn;
exports.quadraticInOut = normalization.quadraticInOut;
exports.quadraticOut = normalization.quadraticOut;
exports.rotate = normalization.rotate;
exports.scale = normalization.scale;
exports.translate = normalization.translate;
exports.validateGraph = normalization.validateGraph;
exports.zIndexOrdering = normalization.zIndexOrdering;
exports.HTML_COLORS = colors.HTML_COLORS;
exports.colorToArray = colors.colorToArray;
exports.colorToIndex = colors.colorToIndex;
exports.extractPixel = colors.extractPixel;
exports.floatColor = colors.floatColor;
exports.getPixelColor = colors.getPixelColor;
exports.indexToColor = colors.indexToColor;
exports.parseColor = colors.parseColor;
exports.rgbaToFloat = colors.rgbaToFloat;
exports.assign = data.assign;
exports.assignDeep = data.assignDeep;
exports.extend = data.extend;
exports.isPlainObject = data.isPlainObject;
